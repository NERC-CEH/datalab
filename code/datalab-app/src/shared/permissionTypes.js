const CREATE = 'create';
const DELETE = 'delete';
const LIST = 'list';
const OPEN = 'open';
const GRANT = 'grant';
const FORBID = 'forbid';

const STACKS = 'stacks';
const STORAGE = 'storage';
const USERS = 'users';

export const PROJECT = 'project';

const keyDelim = '_';
export const permissionDelim = ':';

const elementsPermissionList = {
  CREATE,
  DELETE,
  LIST,
  OPEN,
};

const usersPermissionList = {
  LIST,
  GRANT,
  FORBID,
};

const elements = {
  STACKS,
  STORAGE,
};

const projects = {
  PROJECT,
};

const concatPermissions = (head, tail, char) => `${head}${char}${tail}`;

const makePermissionObj = ([outerKey, outerValue], [innerKey, innerValue]) => ({
  [concatPermissions(outerKey, innerKey, keyDelim)]: concatPermissions(outerValue, innerValue, permissionDelim),
});

const flatMapPermissions = (outer, inner) => Object.entries(outer)
  .map(outerPairs => Object.entries(inner)
    .map(innerPairs => makePermissionObj(outerPairs, innerPairs)))
  .reduce((previous, current) => [...previous, ...current], [])
  .reduce((previous, current) => Object.assign(previous, current), {});

export const elementPermissions = flatMapPermissions(elements, elementsPermissionList);

export const usersPermissions = flatMapPermissions({ USERS }, usersPermissionList);

export const projectPermissions = flatMapPermissions(projects, { ...elementPermissions, ...usersPermissions });

export default {
  ...elementPermissions,
  ...usersPermissions,
  ...projectPermissions,
};
