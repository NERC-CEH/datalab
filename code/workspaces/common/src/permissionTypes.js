const READ = 'read';
const CREATE = 'create';
const DELETE = 'delete';
const LIST = 'list';
const OPEN = 'open';
const EDIT = 'edit';
const ADMIN = 'admin';

const STACKS = 'stacks';
const STORAGE = 'storage';
const SETTINGS = 'settings';
const PERMISSIONS = 'permissions';
const USERS = 'users';
const INSTANCE = 'instance';
const PROJECTS = 'projects';

const PROJECT_NAMESPACE = 'projects';
const KEY_TOKEN = '?projectKey?';
const SYSTEM = 'system';

// key name used for admin role boolean in auth
const INSTANCE_ADMIN_ROLE = 'instanceAdmin';

const keyDelim = '_';
const permissionDelim = ':';

const elementsPermissionList = {
  READ,
  CREATE,
  DELETE,
  LIST,
  OPEN,
  EDIT,
};

const usersPermissionList = {
  LIST,
};

const elements = {
  STACKS,
  STORAGE,
  SETTINGS,
  PERMISSIONS,
  PROJECTS,
};

const PROJECT_KEY = `${PROJECT_NAMESPACE}${permissionDelim}${KEY_TOKEN}`;
const projectKeys = {
  PROJECT_KEY,
};

const concatPermissions = (head, tail, char) => `${head}${char}${tail}`;

const makePermissionObj = ([outerKey, outerValue], [innerKey, innerValue]) => ({
  [concatPermissions(outerKey, innerKey, keyDelim)]: concatPermissions(outerValue, innerValue, permissionDelim),
});

const flatMapPermissions = (outer, inner) => Object.entries(outer)
  .map(outerPairs => Object.entries(inner)
    .map(innerPairs => makePermissionObj(outerPairs, innerPairs)))
  .reduce((previous, current) => [...previous, ...current], [])
  .reduce((previous, current) => Object.assign(previous, current), {});

const elementPermissions = flatMapPermissions(elements, elementsPermissionList);

const usersPermissions = flatMapPermissions({ USERS }, usersPermissionList);

const projectPermissions = flatMapPermissions(projectKeys, { ...elementPermissions, ...usersPermissions });

const projectKeyPermission = (permission, projectKey) => permission.replace(KEY_TOKEN, projectKey);

const systemPermissions = flatMapPermissions({ SYSTEM }, flatMapPermissions({ INSTANCE }, { ADMIN }));

const { SYSTEM_INSTANCE_ADMIN } = systemPermissions;

export {
  INSTANCE_ADMIN_ROLE,
  SYSTEM_INSTANCE_ADMIN,
  PROJECT_NAMESPACE,
  SYSTEM,
  elementPermissions,
  usersPermissions,
  projectPermissions,
  systemPermissions,
  permissionDelim as delimiter,
  projectKeyPermission,
};
