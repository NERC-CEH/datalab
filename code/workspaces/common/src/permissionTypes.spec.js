import { elementPermissions, projectPermissions, systemPermissions } from './permissionTypes';

describe('PermissionTypes', () => {
  it('generates expected permissions', () => expect(elementPermissions).toMatchSnapshot());

  it('generates expected permissions with the project prefix', () => expect(projectPermissions).toMatchSnapshot());

  it('generates expected system permissions', () => expect(systemPermissions).toMatchSnapshot());
});
