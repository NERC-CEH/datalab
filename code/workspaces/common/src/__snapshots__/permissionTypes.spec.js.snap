// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PermissionTypes generates expected permissions 1`] = `
Object {
  "CLUSTERS_CREATE": "clusters:create",
  "CLUSTERS_DELETE": "clusters:delete",
  "CLUSTERS_EDIT": "clusters:edit",
  "CLUSTERS_LIST": "clusters:list",
  "CLUSTERS_OPEN": "clusters:open",
  "CLUSTERS_READ": "clusters:read",
  "PERMISSIONS_CREATE": "permissions:create",
  "PERMISSIONS_DELETE": "permissions:delete",
  "PERMISSIONS_EDIT": "permissions:edit",
  "PERMISSIONS_LIST": "permissions:list",
  "PERMISSIONS_OPEN": "permissions:open",
  "PERMISSIONS_READ": "permissions:read",
  "PROJECTS_CREATE": "projects:create",
  "PROJECTS_DELETE": "projects:delete",
  "PROJECTS_EDIT": "projects:edit",
  "PROJECTS_LIST": "projects:list",
  "PROJECTS_OPEN": "projects:open",
  "PROJECTS_READ": "projects:read",
  "SETTINGS_CREATE": "settings:create",
  "SETTINGS_DELETE": "settings:delete",
  "SETTINGS_EDIT": "settings:edit",
  "SETTINGS_LIST": "settings:list",
  "SETTINGS_OPEN": "settings:open",
  "SETTINGS_READ": "settings:read",
  "STACKS_CREATE": "stacks:create",
  "STACKS_DELETE": "stacks:delete",
  "STACKS_EDIT": "stacks:edit",
  "STACKS_LIST": "stacks:list",
  "STACKS_OPEN": "stacks:open",
  "STACKS_READ": "stacks:read",
  "STORAGE_CREATE": "storage:create",
  "STORAGE_DELETE": "storage:delete",
  "STORAGE_EDIT": "storage:edit",
  "STORAGE_LIST": "storage:list",
  "STORAGE_OPEN": "storage:open",
  "STORAGE_READ": "storage:read",
}
`;

exports[`PermissionTypes generates expected permissions with the project prefix 1`] = `
Object {
  "PROJECT_KEY_CLUSTERS_CREATE": "projects:?projectKey?:clusters:create",
  "PROJECT_KEY_CLUSTERS_DELETE": "projects:?projectKey?:clusters:delete",
  "PROJECT_KEY_CLUSTERS_EDIT": "projects:?projectKey?:clusters:edit",
  "PROJECT_KEY_CLUSTERS_LIST": "projects:?projectKey?:clusters:list",
  "PROJECT_KEY_CLUSTERS_OPEN": "projects:?projectKey?:clusters:open",
  "PROJECT_KEY_CLUSTERS_READ": "projects:?projectKey?:clusters:read",
  "PROJECT_KEY_PERMISSIONS_CREATE": "projects:?projectKey?:permissions:create",
  "PROJECT_KEY_PERMISSIONS_DELETE": "projects:?projectKey?:permissions:delete",
  "PROJECT_KEY_PERMISSIONS_EDIT": "projects:?projectKey?:permissions:edit",
  "PROJECT_KEY_PERMISSIONS_LIST": "projects:?projectKey?:permissions:list",
  "PROJECT_KEY_PERMISSIONS_OPEN": "projects:?projectKey?:permissions:open",
  "PROJECT_KEY_PERMISSIONS_READ": "projects:?projectKey?:permissions:read",
  "PROJECT_KEY_PROJECTS_CREATE": "projects:?projectKey?:projects:create",
  "PROJECT_KEY_PROJECTS_DELETE": "projects:?projectKey?:projects:delete",
  "PROJECT_KEY_PROJECTS_EDIT": "projects:?projectKey?:projects:edit",
  "PROJECT_KEY_PROJECTS_LIST": "projects:?projectKey?:projects:list",
  "PROJECT_KEY_PROJECTS_OPEN": "projects:?projectKey?:projects:open",
  "PROJECT_KEY_PROJECTS_READ": "projects:?projectKey?:projects:read",
  "PROJECT_KEY_SETTINGS_CREATE": "projects:?projectKey?:settings:create",
  "PROJECT_KEY_SETTINGS_DELETE": "projects:?projectKey?:settings:delete",
  "PROJECT_KEY_SETTINGS_EDIT": "projects:?projectKey?:settings:edit",
  "PROJECT_KEY_SETTINGS_LIST": "projects:?projectKey?:settings:list",
  "PROJECT_KEY_SETTINGS_OPEN": "projects:?projectKey?:settings:open",
  "PROJECT_KEY_SETTINGS_READ": "projects:?projectKey?:settings:read",
  "PROJECT_KEY_STACKS_CREATE": "projects:?projectKey?:stacks:create",
  "PROJECT_KEY_STACKS_DELETE": "projects:?projectKey?:stacks:delete",
  "PROJECT_KEY_STACKS_EDIT": "projects:?projectKey?:stacks:edit",
  "PROJECT_KEY_STACKS_LIST": "projects:?projectKey?:stacks:list",
  "PROJECT_KEY_STACKS_OPEN": "projects:?projectKey?:stacks:open",
  "PROJECT_KEY_STACKS_READ": "projects:?projectKey?:stacks:read",
  "PROJECT_KEY_STORAGE_CREATE": "projects:?projectKey?:storage:create",
  "PROJECT_KEY_STORAGE_DELETE": "projects:?projectKey?:storage:delete",
  "PROJECT_KEY_STORAGE_EDIT": "projects:?projectKey?:storage:edit",
  "PROJECT_KEY_STORAGE_LIST": "projects:?projectKey?:storage:list",
  "PROJECT_KEY_STORAGE_OPEN": "projects:?projectKey?:storage:open",
  "PROJECT_KEY_STORAGE_READ": "projects:?projectKey?:storage:read",
  "PROJECT_KEY_USERS_LIST": "projects:?projectKey?:users:list",
}
`;

exports[`PermissionTypes generates expected system permissions 1`] = `
Object {
  "SYSTEM_DATA_MANAGER": "system:data:manager",
  "SYSTEM_INSTANCE_ADMIN": "system:instance:admin",
}
`;
